### PR TITLE
adding get_lifecycle_out_of_recovery clean step to get lifecycle out of recovery

### PR DESCRIPTION
**Pull request by mpardhi23**
@cdearborn @rpioso @digambar15
Added new clean step "get_lifecycle_out_of_recovery" in management interface to take lifecycle controller out of recovery. 
I have tested this changes by executing get_lifecycle_out_of_recovery clean step and also as part of known_good_state clean step.

### CODE REVIEW

File: ironic/drivers/modules/drac/management.py [hunk 1]
```diff
@@ -387,5 +400,81 @@ def known_good_state(self, task):
         node = task.node
 
         client = drac_common.get_drac_client(node)
+        if client.is_lifecycle_in_recovery():
+            get_lcc_out_of_recovery_mode(node)
```
Comment (digambar15) [inline ironic/drivers/modules/drac/management.py]: It will be good to add log.info("Life cycle controller is in recovery mode")

**Comment by mpardhi23** [comment]
I will update the corresponding test cases for this changes , once this code changes get approved.

File: ironic/drivers/modules/drac/management.py [hunk 2]
```diff
@@ -387,5 +400,81 @@ def known_good_state(self, task):
...
+            LOG.info("Lifecycle Controller is out of recovery mode")
+        else:
+            LOG.info("Lifecycle Controller is not in recovery mode")
```
Comment (digambar15) [inline ironic/drivers/modules/drac/management.py]: I suggest to remove else clause

File: ironic/drivers/modules/drac/management.py [hunk 3]
```diff
@@ -387,5 +400,81 @@ def known_good_state(self, task):
...
+    def get_lifecycle_out_of_recovery(self, task):
+        """Take Lifecycle Controller out of recovery mode.
+
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
+        node = task.node
+
+        client = drac_common.get_drac_client(node)
```
Comment (digambar15) [inline ironic/drivers/modules/drac/management.py:421]: ditto as above

File: ironic/drivers/modules/drac/management.py [hunk 4]
```diff
@@ -387,5 +400,81 @@ def known_good_state(self, task):
...
+        if client.is_lifecycle_in_recovery():
+            get_lcc_out_of_recovery_mode(node)
+            LOG.info("Lifecycle Controller is out of recovery mode")
+        else:
```
Comment (digambar15) [inline ironic/drivers/modules/drac/management.py]: Remove this else clause

File: ironic/drivers/modules/drac/management.py (hunk 1 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/management.py]: Sure. Will add this log message.

File: ironic/drivers/modules/drac/management.py (hunk 2 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/management.py]: sure.

File: ironic/drivers/modules/drac/management.py (hunk 4 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/management.py]: ok

File: ironic/conf/drac.py [hunk 5]
```diff
@@ -21,7 +21,14 @@
                min=1,
                help=_('Interval (in seconds) between periodic RAID job status '
                       'checks to determine whether the asynchronous RAID '
-                      'configuration was successfully finished or not.'))
+                      'configuration was successfully finished or not.')),
+    cfg.IntOpt('query_management_config_job_status_interval',
+               default=120,
+               min=1,
+               help=_('Interval (in seconds) between periodic MANAGEMENT job '
```
Comment (cdearborn) [inline ironic/conf/drac.py]: Could you change MANAGEMENT to lower case on the above line and below?

File: ironic/drivers/modules/drac/common.py [hunk 6]
```diff
@@ -115,3 +115,8 @@ def get_drac_client(node):
                                     driver_info['drac_protocol'])
 
     return client
+
+
+def get_reboot_options():
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/common.py]: I think this change may be showing up because I resynced my fork with Ironic.  If so, can you undo this change?

File: ironic/drivers/modules/drac/job.py [hunk 7]
```diff
@@ -72,6 +72,9 @@ def list_unfinished_jobs(node):
     """
     client = drac_common.get_drac_client(node)
 
+    if client.is_lifecycle_in_recovery():
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/job.py:75]: Why do we need this?

File: ironic/conf/drac.py (hunk 5 above)
Comment (mpardhi23) [reply ironic/conf/drac.py]: Sure Chris.

File: ironic/drivers/modules/drac/common.py (hunk 6 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/common.py]: Sure will undo this changes.

File: ironic/drivers/modules/drac/job.py (hunk 7 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/job.py:75]: We have added this statement in list_unfinished_jobs(node) function because,  when client.list_jobs() get invoked then we are getting "timed out error".
As lifecycle is in recovery mode , we are not able to invoke any WSMan API call to get the jobs list.
That's why we have come up with this workaround that is checking if the lifecycle is in recovery, if so then we are returning from that function. 
And the same "if client.is_lifecycle_in_recovery()" statement we have added in get_power_state() and set_power_state() method of power.py also to handle the "timed out error".

File: ironic/drivers/modules/drac/management.py [hunk 8]
```diff
@@ -257,6 +263,13 @@ def set_boot_device(node, device, persistent=False):
         raise exception.DracOperationError(error=exc)
 
 
+def get_lcc_out_of_recovery_mode(node):
+    client = drac_common.get_drac_client(node)
+    settings = {'Lifecycle Controller State': 'Enabled'}
+    client.set_lifecycle_settings(settings)
+    client.commit_pending_lifecycle_changes(reboot=False)
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py:270]: Does this job run without rebooting the server?

File: ironic/drivers/modules/drac/management.py [hunk 9]
```diff
@@ -387,5 +400,77 @@ def known_good_state(self, task):
         node = task.node
 
         client = drac_common.get_drac_client(node)
+        if client.is_lifecycle_in_recovery():
+            LOG.info("Lifecycle Controller is in recovery mode")
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py]: Since ironic may perform this operation on multiple nodes at once, I think we should add some identifier for the node in the above log message.

File: ironic/drivers/modules/drac/management.py [hunk 10]
```diff
@@ -387,5 +400,77 @@ def known_good_state(self, task):
...
+            get_lcc_out_of_recovery_mode(node)
         client.reset_idrac(force=True, wait=True)
         client.delete_jobs(job_ids=[_CLEAR_JOB_IDS])
+
+    @METRICS.timer('DracManagement.get_lifecycle_out_of_recovery')
+    @base.clean_step(priority=0)
+    def get_lifecycle_out_of_recovery(self, task):
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py]: Would you mind changing "get_lifecycle_out_of_recovery" to "take_lifecycle_out_of_recovery"?

File: ironic/drivers/modules/drac/management.py [hunk 11]
```diff
@@ -183,9 +185,13 @@ def set_boot_device(node, device, persistent=False):
     :raises: DracOperationError on an error from python-dracclient.
     """
 
-    drac_job.validate_job_queue(node)
-
     client = drac_common.get_drac_client(node)
+    if client.is_lifecycle_in_recovery():
+        node.driver_internal_info['management_config_substep'] = 'completed'
```
Comment (cdearborn) [inline ironic/drivers/modules/drac/management.py:190]: It looks to me like this code will make the cleaning step never run.  I think this is the flow that will happen:
set_boot_device is called to set the server to PXE boot 1 time
This code sets the cleaning step to completed
The server will not power cycle or PXE boot
The periodic job will trigger and mark cleaning as done

**Review by cdearborn — CHANGES_REQUESTED** [review]
I think we're going to have to discuss what the issues are with this and how to approach them.  It looks to me like the call to take the LCC out of recovery mode will never happen with the current code.
If we can't set the server to PXE boot one time, then the cleaning step will never run unless they have fixed the problem where the server always reboots even when the user is requesting an OOB cleaning step.

File: ironic/drivers/modules/drac/management.py (hunk 8 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/management.py:270]: yes.

File: ironic/drivers/modules/drac/management.py (hunk 10 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/management.py]: Sure, I will change the method name.

File: ironic/drivers/modules/drac/management.py (hunk 9 above)
Comment (mpardhi23) [reply ironic/drivers/modules/drac/management.py]: Sure Chris.

**Comment by mpardhi23** [comment]
Updated required changes as suggested in review comments.

**Comment by cdearborn** [comment]
Hey Monica, I'm going to go ahead and close this PR for now since we will need changes to the Ironic cleaning infrastructure to make this work.  Hopefully we'll be able to make those changes and then revisit this PR at that time.  Thanks for all your hard work on this!
